src/Payments.sol

```diff
@@ -0,0 +1,612 @@
+            require(
+                arbResult.settleUpto <= epochEnd,
+                "failed to settle: arbiter settled beyond segment end"
+            );
+            require(
+                arbResult.modifiedAmount <= rate * (arbResult.settleUpto - epochStart),
+                "failed to settle: arbiter modified amount exceeds maximum for settled duration"
+            );
+
+            // Adjust the amount and the final epoch according to the arbiter's decision.
+            settledUntil = arbResult.settleUpto;
+            settledAmount = arbResult.modifiedAmount;
+            note = arbResult.note;
+        }
+
+        // Revert if no progress was made (i.e., the final epoch did not advance).
+        require(settledUntil > epochStart, "failed to settle: no progress made in settlement");
+
+        return (settledAmount, settledUntil, note);
+    }
+
+    function _settleWithRateChanges(
+        uint256 railId,
+        Rail storage rail,
+        RateChangeQueue.Queue storage rateQueue,
+        uint256 initialRate,
+        uint256 startEpoch,
+        uint256 targetEpoch
+    ) internal returns (uint256 totalSettled, uint256 finalEpoch, string memory note) {
+        totalSettled = 0;
+        uint256 currentEpoch = startEpoch;
+        uint256 activeRate = initialRate;
+        note = "";
+
+        while (currentEpoch <= targetEpoch) {
+            uint256 nextBoundary = targetEpoch;
+
+            // If there's an upcoming rate change, check if it applies within our current range.
+            if (!rateQueue.isEmpty()) {
+                RateChangeQueue.RateChange memory upcomingChange = rateQueue.peek();
+                bool isWithinRange = (
+                    upcomingChange.untilEpoch >= currentEpoch &&
+                    upcomingChange.untilEpoch <= targetEpoch
+                );
+                require(isWithinRange, "rate queue is in an invalid state");
+                nextBoundary = upcomingChange.untilEpoch;
+                activeRate = upcomingChange.rate;
+            }
+
+            // Settle the segment from the current epoch up to the next boundary (or early if arbitration says so).
+            (uint256 segmentAmount, uint256 segmentEndEpoch, string memory arbNote) = _settleSegment(
+                railId,
+                rail,
+                currentEpoch,
+                nextBoundary,
+                activeRate
+            );
+
+            // Update the total settled. 
+            totalSettled += segmentAmount;
+
+            // If the arbitration shortened the segment, stop and return immediately.
+            // but keep the rate change in the queue as we've not fully settled the segment.
+            if (segmentEndEpoch < nextBoundary) {
+                return (totalSettled, segmentEndEpoch, arbNote);
+            }
+
+            // Otherwise, we have fully settled up to nextBoundary.
+            currentEpoch = segmentEndEpoch;
+            note = arbNote;
```

Comment:
It would be kind of nice to somehow join these notes together when non-empty, but that isn't critical. We mostly care about the last note anyways.